packages/zoe/src/contracts/otcDeskReputation.js

```diff
@@ -0,0 +1,142 @@
+// @ts-check
+
+import { E } from '@agoric/eventual-send';
+import { assert, details } from '@agoric/assert';
+import { trade } from '../contractSupport';
+
+import '../../exported';
+
+/**
+ * This OTC Desk contract does not assume that the underlying assets
+ * have been escrowed. Rather, a reputation mechanism
+ * is used to inform potential users of how often the quote is
+ * accurate, and how often the trade fails.
+ *
+ * @type {ContractStartFn}
+ */
+const start = zcf => {
+  const { zcfSeat: marketMakerSeat } = zcf.makeEmptySeatKit();
+
+  let numTradesAttempted = 0;
+  let numTradesDefected = 0;
+
+  const makeQuote = (price, assets, timerAuthority, deadline) => {
+    let quoteActive = true;
+    E(timerAuthority).setWakeup(
+      deadline,
+      harden({ wake: () => (quoteActive = false) }),
+    );
+    const acceptQuote = seat => {
+      assert(quoteActive, details`The quote was no longer active`);
+      const { want: userWantActual, give: userGiveActual } = seat.getProposal();
+      numTradesAttempted += 1;
+      const hasInventory = Object.entries(assets).every(([keyword, amount]) => {
+        const allocatedAmount = marketMakerSeat.getAmountAllocated(
+          keyword,
+          amount.brand,
+        );
+        return zcf.getAmountMath(amount.brand).isGTE(allocatedAmount, amount);
+      });
+      if (!hasInventory) {
+        numTradesDefected += 1;
+        throw Error(
+          `The market maker did not have the inventory they promised. Their rating has been adjusted accordingly`,
+        );
+      }
+
+      // If this trade does not conserve rights, it will throw an
+      // error and the user will get their funds back.
+      try {
+        trade(
```

Comment:
Can this be done in terms of stages and reallocate?  that lets you present the really basic concepts.

Reply:
Ah. That would make the marketMaker's statement more complicated because you'd have to compute their new pool. But that would still be pretty clear. And it would show that we are safely atomic within all that.